feat(core): introduce `BootstrapContext` for improved server bootstrapping

alan-agius4 opened the pull request:
This commit introduces a number of changes to the server bootstrapping process to make it more robust and less error-prone, especially for concurrent requests.

Previously, the server rendering process relied on a module-level global platform injector. This could lead to issues in server-side rendering environments where multiple requests are processed concurrently, as they could inadvertently share or overwrite the global injector state.

The new approach introduces a `BootstrapContext` that is passed to the `bootstrapApplication` function. This context provides a platform reference that is scoped to the individual request, ensuring that each server-side render has an isolated platform injector. This prevents state leakage between concurrent requests and makes the overall process more reliable.

BREAKING CHANGE:
The server-side bootstrapping process has been changed to eliminate the reliance on a global platform injector.

Before:
```ts
const bootstrap = () => bootstrapApplication(AppComponent, config);
```

After:
```ts
const bootstrap = (context: BootstrapContext) =>
  bootstrapApplication(AppComponent, config, context);
```

A schematic is provided to automatically update `main.server.ts` files to pass the `BootstrapContext` to the `bootstrapApplication` call.

In addition, `getPlatform()` and `destroyPlatform()` will now return `null` and be a no-op respectively when running in a server environment.

alan-agius4 commented:
**caretaker note**: this request a separate patch in G3 https://cl/802466606

josephperrott reviewed (approved):
Reviewed-for: dev-infra

alan-agius4 commented on integration/platform-server-hydration/src/main.server.ts:
```diff
@@ -1,7 +1,7 @@
 import {config} from './app/app.config.server';
 
-const bootstrap = () => bootstrapApplication(AppComponent, config);
+const bootstrap = bootstrapServerApplication(AppComponent, config);
 
 export default bootstrap;
```
Yeah we can, although I did not change it in the migration.

alan-agius4 committed:
feat(core): introduce `BootstrapContext` for improved server bootstrapping

This commit introduces a number of changes to the server bootstrapping process to make it more robust and less error-prone, especially for concurrent requests.

Previously, the server rendering process relied on a module-level global platform injector. This could lead to issues in server-side rendering environments where multiple requests are processed concurrently, as they could inadvertently share or overwrite the global injector state.

The new approach introduces a `BootstrapContext` that is passed to the `bootstrapApplication` function. This context provides a platform reference that is scoped to the individual request, ensuring that each server-side render has an isolated platform injector. This prevents state leakage between concurrent requests and makes the overall process more reliable.

BREAKING CHANGE:
The server-side bootstrapping process has been changed to eliminate the reliance on a global platform injector.

Before:
```ts
const bootstrap = () => bootstrapApplication(AppComponent, config);
```

After:
```ts
const bootstrap = (context: BootstrapContext) =>
  bootstrapApplication(AppComponent, config, context);
```

A schematic is provided to automatically update `main.server.ts` files to pass the `BootstrapContext` to the `bootstrapApplication` call.

In addition, `getPlatform()` and `destroyPlatform()` will now return `null` and be a no-op respectively when running in a server environment.

AndrewKushnir commented on packages/core/src/application/create_application.ts:
```diff
@@ -38,17 +40,28 @@ export function internalCreateApplication(config: {
+      RuntimeErrorCode.PLATFORM_NOT_FOUND,
+      ngDevMode &&
+        'Missing Platform: This may be due to using `bootstrapApplication` on the server without passing a `BootstrapContext`. ' +
+          'Please make sure that `bootstrapApplication` is called with a `BootstrapContext.',
```
```suggestion
          'Please make sure that `bootstrapApplication` is called with the `context` argument.',
```

AndrewKushnir commented on packages/core/src/platform/platform.ts:
```diff
@@ -27,30 +27,29 @@ import {PLATFORM_DESTROY_LISTENERS} from './platform_destroy_listeners';
     throw new RuntimeError(
       RuntimeErrorCode.MULTIPLE_PLATFORMS,
-      ngDevMode && 'There can be only one platform. Destroy the previous one to create a new one.',
+      ngDevMode &&
+        ngDevMode &&
```
Is this a typo (2 `ngDevMode`s) or the idea was to use `ngServerMode` here?

AndrewKushnir commented on packages/core/src/platform/platform.ts:
```diff
@@ -135,16 +133,26 @@ export function assertPlatform(requiredToken: any): PlatformRef {
 /**
  * Returns the current platform.
```
nit: rephrased a bit and suggested an alternative way of obtaining this info:
```suggestion
 * Returns the current platform in the browser environment. In the server environment,
 * returns `null`. If you need access to the platform info, inject `PlatformRef` in your application
 * code.
```

AndrewKushnir replied:
If we use this content, we can remove the `@remarks` part.

alan-agius4 replied:
I removed it, although In general, I do prefer to use `@remarks` to callout important information.

AndrewKushnir reviewed (approved):
LGTM, just a few minor comments.

alan-agius4 committed:
fixup! feat(core): introduce `BootstrapContext` for improved server bootstrapping

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

hybrist reviewed (approved):
Reviewed-for: public-api
